### ISSUE TITLE
no-undef should apply on assigning a property to an undefined variable

### ISSUE THREAD

**Issue by hexagon6** [eslint/eslint#11840]
**What rule do you want to change?**
no-undef

**Does this change cause the rule to produce more or fewer warnings?**
more

**How will the change be implemented? (New option, new default behavior, etc.)?**
maybe a new option

**Please provide some example code that this change will affect:**
The following assignment should not be allowed if the rule is active, assuming `Text` is an undefined variable.

```js
Text.propTypes = { a: 'b' }
```

**What does the rule currently do for this code?**
It does not apply on this case and can potentially break builds.

**What will the rule do after it's changed?**
It will warn or error if there is an property-assignment to an undefined variable.

**Are you willing to submit a pull request to implement this change?**
Yes

**Comment by g-plane** [comment]
I can't reproduce：https://eslint.org/demo/#eyJ0ZXh0IjoiVGV4dC5wcm9wVHlwZXMgPSB7IGE6ICdiJyB9Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo5LCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLXVuZGVmIjoyfSwiZW52Ijp7fX19

**Comment by hexagon6** [comment]
@g-plane It seems this option does not error when the environment browser is enabled.

Is Text an allowed browser global?

**Comment by aladdin-add** [comment]
> Is Text an allowed browser global?

seems it was true: https://github.com/sindresorhus/globals/blob/8f022be1112cc0964ca77947a4db5417056188c8/globals.json#L874

so it was working as expected. :-)

**Comment by hexagon6** [comment]
Great! Thank you for helping out. I think it can be closed. Was surprised that this does not fail.

**Comment by hexagon6** [comment]
Funny :-) https://dom.spec.whatwg.org/#interface-text
